nameko_grpc/errors.py

```diff
@@ -1,38 +1,133 @@
     def from_headers(headers):
-        status = int(headers.get("grpc-status"))
+        """ Rehydrate a new instance from headers received as trailing metadata.
+        """
+        code = int(headers.get("grpc-status"))
         message = headers.get("grpc-message")
+        status = headers.get(GRPC_DETAILS_METADATA_KEY)
 
         return GrpcError(
-            status=CYGRPC_STATUS_CODE_TO_STATUS_CODE[status], details=message
+            code=STATUS_CODE_INT_TO_ENUM_MAP[code],
+            message=message,
+            status=Status.FromString(status) if status else None,
         )
 
+    @staticmethod
+    def from_status(status):
+        return GrpcError(
+            code=STATUS_CODE_INT_TO_ENUM_MAP[status.code],
+            message=status.message,
+            status=status,
+        )
+
+    @staticmethod
+    def from_exception(exc_info, code=None, message=None):
+        """ Utility method to create a new GrpcError instance representing an
+        underlying exception. Useful in try/except clauses.
+
+        By default, a `google.rpc.Status` message will be generated capturing the
+        debug info of the underyling traceback. See `default_error_from_exception`.
+
+        This can be overridden by registering a new callable against a given exception
+        type. See `register_exception_handler`.
+        """
+        exc_type, exc, tb = exc_info
+
+        error_from_exception = registry.get(exc_type, default_error_from_exception)
+        return error_from_exception(exc_info, code, message)
+
     def __str__(self):
         return (
             "<RPC terminated with:\n"
-            "\tstatus = {}\n"
-            '\tdetails = "{}"\n'
-            '\tdebug_error_string = "{}"\n'
-            ">".format(self.status, self.details, self.debug_error_string)
+            "\tcode = {}\n"
+            '\tmessage = "{}"\n'
+            '\tstatus = "{}"\n'
+            ">".format(self.code, self.message, self.status)
+        )
+
+
+def make_status(code, message, details=None):
+    return Status(code=code.value[0], message=message, details=details or [])
+
+
+def register_exception_handler(exc_type, custom_error_from_exception):
+    """ Register a custom implementation to generate a GrpcError from an underlying
+    exception, by exception type.
+
+    Must be a callable with a signature matching `default_error_from_exception`.
+    """
+    registry[exc_type] = custom_error_from_exception
+
+
+def unregister_expection_handler(exc_type):
+    """ Unregister a custom implementation.
+    """
+    registry.pop(exc_type, None)
+
+
+def default_error_from_exception(exc_info, code=None, message=None):
+    """ Create a new GrpcError instance representing an underlying exception.
+
+    If the `DEBUG` key is set in the Nameko config, the `status` message will capture
+    the underyling traceback in a `google.rpc.error_details.DebugInfo` message.
+    """
+    exc_type, exc, tb = exc_info
+
+    code = code or StatusCode.UNKNOWN
```

Comment:
should UNKNOWN be the default unhandled error code? or should it be INTERNAL? Maybe it does not matter much .)

Reply:
The default of UNKNOWN follows what the official Python gRPC client does.

Reply:
:+1: